Fuzzlyn/Methods/FuncBodyGenerator.cs

```diff
@@ -342,37 +358,106 @@ private StatementSyntax GenReturn()
-                                SyntaxKind.NumericLiteralExpression,
-                                Literal(0))))));
+                            lowerBound))));
 
         ExpressionSyntax cond =
             BinaryExpression(
-                SyntaxKind.LessThanExpression,
+                endCondition,
                 IdentifierName(varName),
-                LiteralExpression(
-                    SyntaxKind.NumericLiteralExpression,
-                    Literal(2)));
+                upperBound);
 
         ExpressionSyntax incr =
-            PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, indVar.Expression);
+            PostfixUnaryExpression(nextValueExpression, indexVar.Expression);
 
-        BlockSyntax block = GenBlock([indVar]);
+        BlockSyntax block = GenBlock([indexVar]);
 
         ForStatementSyntax @for = ForStatement(decl, SeparatedList<ExpressionSyntax>(), cond, SingletonSeparatedList(incr), block);
         return @for;
     }
 
+    private StatementSyntax GenDoLoop()
+    {
+        string varName = $"lvar{_varCounter++}";
+        SyntaxKind op = (SyntaxKind)Options.LoopIndexTypeDist.Sample(_random.Rng);
+        PrimitiveType indexPrimType = _types.GetPrimitiveType(op);
+        ScopeValue indexVar = new(indexPrimType, IdentifierName(varName), -(_scope.Count - 1), true);
+        (ExpressionSyntax lowerBound, ExpressionSyntax upperBound) = LiteralGenerator.GenPrimitiveLiteralLoopBounds(_random, indexPrimType);
+
+        bool upCountedLoop = _random.FlipCoin(_random.Options.UpCountedLoopProb);
+        SyntaxKind endCondition = upCountedLoop ? SyntaxKind.LessThanExpression : SyntaxKind.GreaterThanExpression;
+        SyntaxKind nextValueExpression = upCountedLoop ? SyntaxKind.PostIncrementExpression : SyntaxKind.PostDecrementExpression;
+        if (!upCountedLoop)
+        {
+            (lowerBound, upperBound) = (upperBound, lowerBound); // swap the bounds
+        }
+
+        VariableDeclarationSyntax decl =
+            VariableDeclaration(
+                indexVar.Type.GenReferenceTo(),
+                SingletonSeparatedList(
+                    VariableDeclarator(varName)
+                    .WithInitializer(
+                        EqualsValueClause(
+                            lowerBound))));
+
+        ExpressionSyntax cond =
+            BinaryExpression(
+                endCondition,
+                IdentifierName(varName),
+                upperBound);
+
+        ExpressionSyntax incr =
+            PostfixUnaryExpression(nextValueExpression, indexVar.Expression);
+
+        BlockSyntax innerBlock = GenBlock(epilogStatements: [ExpressionStatement(incr)]);
+
+        DoStatementSyntax @do = DoStatement(
+            Token(SyntaxKind.DoKeyword),
+            innerBlock,
+            Token(SyntaxKind.WhileKeyword),
+            Token(SyntaxKind.OpenParenToken),
+            cond,
+            Token(SyntaxKind.CloseParenToken),
+            Token(SyntaxKind.SemicolonToken));
```

Comment:
```suggestion
        DoStatementSyntax @do = DoStatement(innerBlock, cond);
```

I do not think these explicit tokens are necessary

Reply:
I tend to use "Go to definition" in VS to find these -- typically via IL disassembly? -- and the Do form didn't have the same flexibility as the For form, but I'll try it.

Reply:
https://roslynquoter.azurewebsites.net/ is invaluable for this stuff.

Reply:
> https://roslynquoter.azurewebsites.net/ is invaluable for this stuff.

Where is the "mind blown" emoji when you need it?!